### PR TITLE
Upload HeapDump to S3

### PR DESCRIPTION
**Pull request by t-izbassar**
Fix for #400.

Added skeleton classes with relevant puzzles.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#738](https://github.com/zerocracy/farm/issues/738) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#738](https://github.com/zerocracy/farm/issues/738) is assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/651923039). The budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4). Please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

File: src/main/java/com/zerocracy/farm/S3Dump.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import java.nio.file.Path;
+
+/**
+ * Heap dump in S3.
+ *
+ * @author Izbassar Tolegen (t.izbassar@gmail.com)
+ * @version $Id$
+ * @since 1.0
+ */
+@SuppressWarnings({"PMD.SingularField", "PMD.AvoidDuplicateLiterals",
```
Comment (carlosmiranda) [inline src/main/java/com/zerocracy/farm/S3Dump.java]: @t-izbassar we can avoid `PMD.AvoidDuplicateLiterals` by a simple change, see my comments below

File: src/main/java/com/zerocracy/farm/S3Dump.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+     * Heap dump content from S3.
+     * @return Content
+     * @todo #400:30min Should read content from S3 bucket,
+     *  but only if the current storage content is outdated.
+     *  If it is not outdated then return from the storage.
+     *  Should be covered with unit and integration tests.
+     */
+    public InputStream content() {
+        throw new UnsupportedOperationException(
+            "Not yet implemented"
```
Comment (carlosmiranda) [inline src/main/java/com/zerocracy/farm/S3Dump.java]: @t-izbassar let's make this say:

> S3Dump#content() not yet implemented

File: src/main/java/com/zerocracy/farm/S3Dump.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    /**
+     * Upload dump to S3.
+     * @param item What to upload
+     * @todo #400:30min Should upload heap dump to S3 bucket and update storage.
+     *  Should be covered with unit and integration tests.
+     */
+    public void upload(final OutputStream item) {
+        throw new UnsupportedOperationException(
+            "Not yet implemented"
```
Comment (carlosmiranda) [inline src/main/java/com/zerocracy/farm/S3Dump.java]: @t-izbassar same here, we could say

> S3Dump#upload() not yet implemented

**Comment by t-izbassar** [comment]
@carlosmiranda fixed.

**Comment by yegor256** [comment]
@t-izbassar I don't see why we need this `S3Dump` class. What it's for? We will just upload the file to S3 in `Dumps` and show it to the user in `TkDump`. There are no other classes required, I believe. It's over-design.

**Comment by t-izbassar** [comment]
@yegor256 it's encapsulate S3 path and operations on that. With that we won't duplicate that in `Dumps` and `TkDump` and will just use this class with provided content to upload and read content from that. Also it will be responsible to checking if we already have newest version on S3, maintain local temp storage of that and things like that.

But I can delete that if you want.

**Comment by yegor256** [comment]
@t-izbassar it's good to encapsulate all that, but I would still prefer to have one class. `Dumps` or `S3Dump`. Two is too much for this supplementary functionality. Call it `HeapDump` and make it work like that:

```
HeapDump dump = new HeadDump();
// If the file exists, uploads it to S3.
// Ignore it if it's the same as we already have in S3.
dump.save(); 
// Downloads the latest version from S3
InputStream stream = dump.load(); 
```

Then, someone has to call this `save()` method regularly... I would combine this with the problem explained here: #680 We will need some crontab-like system, which should run certain routines regularly. This "heap dump uploading" routine is a perfect candidate.

**Comment by t-izbassar** [comment]
@yegor256 I've deleted `Dumps` and updated methods and puzzles.

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#738](https://github.com/zerocracy/farm/issues/738) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256)

**Comment by ypshenychka** [comment]
@carlosmiranda According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only two issues were found during code review. 
Please confirm that you'll try to find at least three problems while future reviews.

**Comment by carlosmiranda** [comment]
@ypshenychka yes, confirmed

**Comment by ypshenychka** [comment]
@carlosmiranda thanks!

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 points just awarded to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda)

**Comment by 0crat** [comment]
Quality review completed: +8 points just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**